Store sparse checkpoints in database

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#769 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have enabled cascade deletion in SQLite and therefore, let the database cleanup entries instead of doing ourselves.

- [x] I have reviewed how `purgeCheckpoints` works and now instead computes a list of checkpoints that _should be_ in the database. This list is a sparse list that will keep a checkpoint every `100` blocks, as well as `10` checkpoints close to the tip, all of that, making sure that there's no checkpoint older than k. Then, with cascade deletion, purging checkpoints becomes really easy: delete any checkpoints that is not in the list.

- [x] Fixed broken restoration benchmarks: `restoreWallet` call was missing, so the wallet was never getting restored....

# Comments

<!-- Additional comments or screenshots to attach if any -->

- See comment on `sparseCheckpoints` for some examples / details on how it works.

## Restoration benchmarks with this approach:

```
All results:                            
  restore testnet seq: 5.242 s          
  restore testnet 10% ownership: 5.480 s

All results:                            
  restore mainnet seq: 67.72s
  restore testnet 10% ownership: ?? 
```

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
define helper to compute sparse checkpoints refs with corresponding properties

KtorZ committed:
add foreign keys to 'Wallet' & 'Checkpoint' to leverage cascade deletions on rollbacks and purges

KtorZ committed:
leverage cascade deletes in the SQL database

This simplifies quite a lot our Sqlite implementation; SQLite already supports well
cascade deletion, so it'd be very unfortunate to re-implement all the logic ourselves...

KtorZ committed:
move 'sparseCheckpoints' one level up and use it in the wallet layer to avoid overloading the database

KtorZ committed:
fix broken http-bridge restore benchmark: restoration worker wasn't started anymore

KtorZ committed:
re-generate nix machinery

KtorZ committed:
avoid magic numbers: use named constant in sparseCheckpoints

This makes for better readability.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/Properties.hs:
```diff
@@ -217,6 +226,39 @@ properties = do
+                , 2500 , 2600 , 2700 , 2704 , 2705
+                , 2706 , 2707 , 2708 , 2709 , 2710
+                , 2711 , 2712 , 2713 , 2714
+                ]
```
I put some hand-written tests here to really get convinced of the behavior, in addition with the properties below.

KtorZ commented on stack.yaml:
```diff
@@ -18,6 +18,16 @@ extra-deps:
+- git: https://github.com/KtorZ/persistent
+  commit: 79f2ece07eafae005a703c8eda1bd2420b5e07b5
+  subdirs:
+      - persistent-sqlite
```
I originally wanted to fork `persistent` on `input-output-hk` but there's already a fork there that is archived.
I've asked for permission to unarchive it and get push rights on it.

Note that, in the future, we might want to do something less hacky than my hack above. There's some work to follow on the matter here:

https://github.com/yesodweb/persistent/pull/912

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -72,8 +75,7 @@ PrivateKey                             sql=private_key
     Primary privateKeyWalletId
-    Foreign Wallet fk_wallet_private_key privateKeyWalletId
-
+    Foreign Wallet fk_wallet_private_key privateKeyWalletId ! ON DELETE CASCADE
```
Somewhat an ugly / pretty hack. But it works :no_mouth:

paweljakubas commented on lib/core/src/Cardano/Wallet/DB.hs:
```diff
@@ -183,3 +193,69 @@ newtype PrimaryKey key = PrimaryKey key
+        k = fromIntegral $ getQuantity epochStability
+        h = fromIntegral $ getQuantity $ blockHeight blkH
+
+        minH = if h < k + gapsSize then 0 else h - k - gapsSize
```
:100:

paweljakubas reviewed (approved):
LGTM! I am fond of tests added, also unit tests showing explicitly the case. The beautiful picture somewhat spoils the fact that we can restore an empty wallet in 4s on testnet (14 epochs), but it takes 14minutes on mainnet (150 epochs) :thinking:

paweljakubas commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -559,37 +562,30 @@ restoreBlocks ctx wid blocks nodeTip = do
-        let meta' = calculateMetadata (view #slotId $ currentTip cpLast)
-        DB.putCheckpoint db (PrimaryKey wid) cpLast
+        let unstable = sparseCheckpoints k nodeTip
+        forM_  cps $ \cp -> do
```
hypothesis : this nonlinearity when we deal with additional epoch comes from this `cps/sparseCheckpoint` - we are filtering every round from 0 instead of k. So as there is more stuff to digest more is to be done...

KtorZ committed:
review sparseCheckpoints implementation to avoid creating huge thunks when h is big

KtorZ committed:
review arbitrary generators for SlotId and BlockHeader

KtorZ committed:
use blockheight directly as an argument for 'sparseCheckpoints' instead of a full BlockHeader

KtorZ committed:
prune checkpoints based on the most recent one

KtorZ committed:
Make database pruning a separate operation from 'putCheckpoints'

This makes it much easier to test as pruning stop intereferring with
the state-machine and other tests. So we are able to generate arbitrary
block heights and do better testing.
